### ISSUE TITLE
Unsafe snapshots mechanism (location overwriting)

### ISSUE THREAD

**Issue by d0znpp** [elastic/elasticsearch#21580]
**Elasticsearch version**: 5

**Plugins installed**: []

**JVM version**: 7se

**OS version**: Linux Debian 8.5

**Description of the problem including expected versus actual behavior**: it's possible to use the same location to different snapshots.

**Steps to reproduce**:
 1. Obtain the list of snapshots location http://localhost:9200/_snapshot/_all
 2. Create new snapshot with the same location of one of the existing snapshots
 3. See the results. It will overwrite previous snapshot.

**Provide logs (if relevant)**:

**Describe the feature**:

**Comment by abeyad** [comment]
@d0znpp Could you kindly provide more detailed instructions of what you are seeing?  I tried following what I could understand from your instructions by creating a snapshot with the name `foo`, listed my snapshots, then created another snapshot with the name `bar`, and all worked fine.  I also tried creating another snapshot with the name `foo`, and the system correctly prevented me from doing so with the exception that a snapshot by the same name already exists.

**Comment by d0znpp** [comment]
Creating the snapshot 1.
```
PUT /_snapshot/back1 HTTP/1.1
Host: localhost:9200
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 105

{
    "type": "fs",
    "settings": {
        "location": "aaa",
        "compress": true
    }
}
```
```
PUT /_snapshot/back1/xxx HTTP/1.1
Host: localhost:9200
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 90

{
  "indices": "i1",
  "ignore_unavailable": true,
  "include_global_state": false
}
```

Creating the snapshot 2 with the same location (aaa)
```
PUT /_snapshot/back2 HTTP/1.1
Host: localhost:9200
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 105

{
    "type": "fs",
    "settings": {
        "location": "aaa",
        "compress": true
    }
}
```
```
PUT /_snapshot/back2/yyy HTTP/1.1
Host: localhost:9200
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 90

{
  "indices": "i2",
  "ignore_unavailable": true,
  "include_global_state": false
}
```

Then try to restore snapshot 1.

**Comment by abeyad** [comment]
I followed your exact instructions and had no problem restoring the snapshot.

What was the error you got from trying to restore?  Was it by chance 
```cannot restore index [your_index_name] because it's open```

**Comment by d0znpp** [comment]
No error. But the data will be from the snapshot 2 with other indexes. Should be the error

**Comment by abeyad** [comment]
Sorry, I still don't see the issue.  I did the following:

1. created an index of 100 documents
2. took a snapshot `foo` in repository 1
3. indexed 100 more documents
4. took another snapshot `bar` in repository 2 (which points to the same location as repository 1)
5. deleted the index
6. restored the index from snapshot `foo`
7. ran a search - I only got 100 documents as expected, not 200 had the 2nd snapshot been restored.
